core/services/workflows/syncer/workflow_registry.go

```diff
@@ -353,13 +386,190 @@ func (w *workflowRegistry) readRegistryEvents(ctx context.Context, don capabilit
+	engineKeyMap := map[string]bool{}
+
+	for _, wfMeta := range workflowMetadata {
+		// TODO: ensure that the WorkflowRegisteredEvent clears the engine registry as the very last step
+		engine, getErr := w.engineRegistry.Get(EngineRegistryKey{Owner: wfMeta.Owner, Name: wfMeta.WorkflowName})
+		currWfID := hex.EncodeToString(wfMeta.WorkflowID[:])
+
+		// if there is no engine in the engine registry
+		// then handle as registered event
+		if getErr != nil {
+			toRegisteredEvent := WorkflowRegistryWorkflowRegisteredV1{
+				WorkflowID:    wfMeta.WorkflowID,
+				WorkflowOwner: wfMeta.Owner,
+				DonID:         wfMeta.DonID,
+				Status:        wfMeta.Status,
+				WorkflowName:  wfMeta.WorkflowName,
+				BinaryURL:     wfMeta.BinaryURL,
+				ConfigURL:     wfMeta.ConfigURL,
+				SecretsURL:    wfMeta.SecretsURL,
+			}
+			events = append(events, workflowAsEvent{
+				Data:      toRegisteredEvent,
+				EventType: WorkflowRegisteredEvent,
+			})
+			engineKeyMap[currWfID] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, but the metadata has changed
+		// then handle as updated event
+		prevWfID := hex.EncodeToString(engine.workflowID[:])
+		if currWfID != prevWfID {
+			toUpdatedEvent := WorkflowRegistryWorkflowUpdatedV1{
+				OldWorkflowID: engine.workflowID,
+				NewWorkflowID: wfMeta.WorkflowID,
+				WorkflowOwner: wfMeta.Owner,
+				DonID:         wfMeta.DonID,
+				WorkflowName:  wfMeta.WorkflowName,
+				BinaryURL:     wfMeta.BinaryURL,
+				ConfigURL:     wfMeta.ConfigURL,
+				SecretsURL:    wfMeta.SecretsURL,
+			}
+			events = append(events, workflowAsEvent{
+				Data:      toUpdatedEvent,
+				EventType: WorkflowUpdatedEvent,
+			})
+			engineKeyMap[currWfID] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, the metadata is the same, and the status is active
+		// then the workflow is unchanged. mark as seen.
+		if WorkflowStatus(wfMeta.Status) == WorkflowStatusActive {
+			engineKeyMap[wfMeta.WorkflowName+hex.EncodeToString(wfMeta.Owner)] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, the metadata is the same, but the status isn't active
+		// then the workflow has been paused. NOOP - to be handled below as a deleted event, which clears the DB workflow spec.
+	}
+
+	// Shut down engines that are no longer in the contract's latest workflow metadata state
+	allEngines := w.engineRegistry.GetAll()
+	for _, engine := range allEngines {
+		prevWfID := hex.EncodeToString(engine.workflowID[:])
+		_, exists := engineKeyMap[prevWfID]
+		if !exists {
+			toDeletedEvent := WorkflowRegistryWorkflowDeletedV1{
+				WorkflowID:    engine.workflowID,
+				WorkflowOwner: engine.workflowOwner,
+				DonID:         donID,
+				WorkflowName:  engine.workflowName,
+			}
+			events = append(events, workflowAsEvent{
+				Data:      toDeletedEvent,
+				EventType: WorkflowDeletedEvent,
+			})
+		}
+	}
+
+	for _, event := range events {
```

Comment:
_nit_ I would make this function only return a slice and leave the channel writing to the functions called `sync*`.  two benefits:
1. Right now this doesn't listen for `<-ctx.Done()`
2. simplify the signature of this method and make it more easily testable

Reply:
I agree with point 2 that it might make it easier to test. Though setting up a channel isn't too bad.

For point 1, if the `sync*` function instead called `workflowMetadataToEvents` to get events and then sent them down the channel itself, I don't see how that would have any change to respect ctx.Done() mid processing. You would need to check ctx.Done() before each `w.eventCh <- event`, which seems overkill.

Reply:
You will need to always check the context before any writes. 

IMO it's helpful for maintenance to have channel writes happen in as few places as possible.